assets/pages/new-contributions-landing/components/StripePaymentRequestButton.jsx

```diff
@@ -0,0 +1,198 @@
+});
+
+
+// ----- Functions -----//
+
+
+function updateUserEmail(data: Object, setEmail: string => void) {
+  const email = data.payerEmail;
+  if (email) {
+    if (isValidEmail(email)) {
+      setEmail(email);
+    } else {
+      logException(`Failed to set email for stripe payment request user with email: ${email}`);
+    }
+  } else {
+    logException('Failed to set email: no email in data object');
+  }
+}
+
+const onComplete = (complete: Function) => (res: PaymentResult) => {
+  if (res.paymentStatus === 'success') {
+    complete('success');
+  } else if (res.paymentStatus === 'failure') {
+    complete('fail');
+  }
+};
+
+function onClick(event, props: PropTypes) {
+  event.preventDefault();
+  props.setStripePaymentRequestButtonClicked();
+  const amountIsValid =
+    checkAmountOrOtherAmount(
+      props.selectedAmounts,
+      props.otherAmounts,
+      props.contributionType,
+      props.countryGroupId,
+    );
+  if (props.paymentRequest && amountIsValid) {
+    props.paymentRequest.show();
+  }
+}
+
+function initialisePaymentRequest(props: PropTypes) {
+  const paymentRequest = props.stripe.paymentRequest({
+    country: props.country,
+    currency: props.currency,
+    total: {
+      label: 'The Guardian',
+      amount: props.amount,
+    },
+    requestPayerEmail: true,
+  });
+  paymentRequest.on('token', ({ complete, token, ...data }) => {
+    // We need to do this so that we can offer marketing permissions on the thank you page
+    updateUserEmail(data, props.updateEmail);
+    const tokenId = props.isTestUser ? 'tok_visa' : token.id;
+    props.onPaymentAuthorised({ paymentMethod: 'Stripe', token: tokenId }, onComplete(complete));
+  });
+
+  // The returned value from canMakePayment will either be:
+  // . null - browser has no compatible payment method)
+  // . {applePay: true} - applePay is available
+  // . {applePay: false} - GooglePay or PaymentRequestApi available
+  paymentRequest.canMakePayment().then((result) => {
+    if (result && result.applePay === true) {
+      props.setCanMakeApplePayPayment(true);
+    }
+  });
+  props.setPaymentRequest(paymentRequest);
+}
+
+const paymentButtonStyle = {
+  paymentRequestButton: {
+    theme: 'dark',
+    height: '42px',
+  },
+};
+
+function updateAmount(amount: number, paymentRequest: Object) {
+  // When the other tab is clicked, the value of amount is NaN
+  if (!Number.isNaN(amount) && paymentRequest) {
+    paymentRequest.update({
+      total: {
+        label: 'The Guardian',
+        amount: amount * 100,
+      },
+    });
+  }
+}
+
+
+// ---- Component ----- //
+function PaymentRequestButton(props: PropTypes) {
+  // If we haven't initialised the payment request, initialise it and return null, as we can't insert the button
+  // until the async canMakePayment() function has been called on the paymentRequest object.
+  if (!props.paymentRequest) {
+    initialisePaymentRequest({ ...props });
+    return null;
```

Comment:
It's quite confusing how there's one `return null;` in the container element, and then some more checks here. Is it possible to put them all in one place?

Reply:
Or if not, a comment explaining why

Reply:
Before we try and load the Stripe button, we need to check whether the user is in the right country/Stripe has loaded (you can't insert the `<StripeProvider>` tag until it's loaded)/they are in the test. These checks happen at the container level.

Then, once we've decided to load it, we need to check whether a) the `paymentRequest` object has been initialised, and b) whether the browser check has come back true.

So I think the way it currently is laid out makes sense. Basically, loading the `<StripeProvider>` is dependent on some state, and then you can't initialise the `paymentRequest` object until you've injected the `stripe` object into the button (see line 211 in StripePaymentRequestButton), which you get from `StripeProvider`.